nvaccess/nvda
# Fix reporting time in foobar2000 over 24 hours

### Link to issue number:
Fixes #14127
Supersedes #14451

### Summary of the issue:
When using the foobar2000 audio player, the `ctrl+shift+r` command should report the remaining time in the audio file in the SS, MM:SS, HH:MM:SS, or D HH:MM:SS format depending on the remaining time length. 
However, using the command on audio files over a day long produces an error as the code attempts to read the time string in the HH:MM:SS format instead of the D HH:MM:SS format.

### Description of user facing changes
Time calculation is now processed correctly for all audio length up to 31 days. After this point, it is uncertain how foobar reports time. 

In addition to fixing the time calculations, the spoken UI message was also amended to say "[time] total", "[time] elapsed", or "[time] remaining", rather than just "[time]" so as to be more understandable for the listener.

### Description of development approach
Created a collection of translatable time string output formats.

Created a generic way to parse timedeltas to localised formatted strings.

Created internal functions for foobar appModule to parse foobar time strings to localised formatted strings.

### Testing strategy:
- [x] Manual testing (need confirmation):
1. Open foobar2000
2. Open audio file over 24 hours long
3. Skip to multiple time points throughout audio file to test out following steps 4-6:
4. Type ctrl + shift + e and confirm that correct elapsed time is reported
5. Type ctrl + shift + r and confirm that correct remaining time is reported
6. Type ctrl + shift + t and confirm that correct total time is reported
7. Test out commands with audio files under 24 hours long as well

Unit testing has been added

### Known issues with pull request:
It is uncertain how reporting works for files over a month long.

### Change log entries:
```t2t
Bug fixes
- Remaining, elapsed and total time is now reported correctly for audio files over a day long in foobar2000. (#14127)

API breaking changes
- The following symbols have been removed from ``appModules.foobar2000`` with no direct replacement.
  - ``statusBarTimes``
  - ``parseIntervalToTimestamp``
  - ``getOutputFormat``
  - ``getParsingFormat``
  -
```

### Code Review Checklist:



  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - [ ] After merging: Announce API breaking change on add-on API mailing list
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
